DoSomething/gladiator
# Refactoring exports and leaderboard processes to simplify and DRY code up.

#### What's this PR do?
This PR updates the CSV exporting functionality and the leaderboard generation. It DRY's up the code and cleans up the exporting process to make it easier to troubleshoot issues. For the leaderboard it now appends retrieved user reportback activity to the user object for easier retrieval. Also, the leaderboard export functionality was moved to the leaderboard page to help avoid having to make the same series of costly API and data requests when viewing the leaderboard page, and then again when exporting the leaderboard. Instead, now when the leaderboard page is loaded, the data is flashed in cache (1 request cycle) so when exporting the data, it will grab it from that flashed cache session. This also ensures that the leaderboard you see is the one that gets exported.

The leaderboard view also now includes an additional table with the list of contestants that have yet to reportback.

#### Screenshots
![pasted image at 2016_04_18 04_02 pm](https://cloud.githubusercontent.com/assets/105849/14618990/4245b408-0583-11e6-9743-15f4d951a7e8.png)

#### How should this be manually tested?
Pull down, do the usual composer-y thngs. Load a competition and view the leaderboard. Then try exporting the data. Also try exporting the list of competition users on the competitions page and waiting room pages to make sure you get an accurate CSV exported.

#### Any background context you want to provide?
It was a bit difficult to test the retrieval of signup activity data from the set of test users I was using, since there wasn't necessarily data associated with that user and the randomly picked campaign they were assigned to in the seeding process. A lot of manual effort needed to be made to ensure obtaining users that could be tested. There is some concern how the data will be when testing this PR on QA environments.

#### What are the relevant tickets?
Fixes #205

---
@DoSomething/gladiator
